src/compiler/cse.rs

```diff
@@ -0,0 +1,452 @@
+/// queriables. Using the Schwartz-Zippel lemma, we can determine if two expressions are equivalent
+/// with high probability.
+#[allow(dead_code)]
+pub(super) fn cse<F: Field + Hash, S: Scoring<F>>(
+    mut circuit: SBPIRMachine<F, NullTraceGenerator>,
+    config: CseConfig,
+    scorer: &S,
+) -> SBPIRMachine<F, NullTraceGenerator> {
+    for (_, step_type) in circuit.step_types.iter_mut() {
+        cse_for_step(step_type, &circuit.forward_signals, &config, scorer)
+    }
+    circuit
+}
+
+fn cse_for_step<F: Field + Hash, S: Scoring<F>>(
+    step_type: &mut StepType<F, ()>,
+    forward_signals: &[ForwardSignal],
+    config: &CseConfig,
+    scorer: &S,
+) {
+    let mut signal_factory = SignalFactory::default();
+    let mut replaced_hashes = HashSet::new();
+
+    for _ in 0..config.max_iterations {
+        let mut queriables = Vec::<Queriable<F>>::new();
+
+        forward_signals.iter().for_each(|signal| {
+            queriables.push(Queriable::Forward(*signal, false));
+            queriables.push(Queriable::Forward(*signal, true));
+        });
+        step_type.signals.iter().for_each(|signal| {
+            queriables.push(Queriable::Internal(*signal));
+        });
+
+        // Generate random assignments for the queriables
+        let mut rng = ChaCha20Rng::seed_from_u64(0);
+        let random_assignments: VarAssignments<F, Queriable<F>> = queriables
+            .iter()
+            .cloned()
+            .map(|q| (q, F::random(&mut rng)))
+            .collect();
+
+        // Turn all Expr<F, V, ()> into Expr<F, V, HashResult>
+        let mut step_type_with_hash = step_type.transform_meta(|expr| {
+            let hashed_expr = expr.hash(&random_assignments);
+            hashed_expr.meta().clone()
+        });
+
+        // Extract all the expressions from the step type
+        let mut exprs = Vec::<Expr<F, Queriable<F>, HashResult>>::new();
+
+        for constraint in &step_type_with_hash.constraints {
+            exprs.push(constraint.expr.clone());
+        }
+
+        // Find the optimal subexpression to replace
+        if let Some(common_expr) = find_optimal_subexpression(&exprs, &replaced_hashes, scorer) {
+            // Add the hash of the replaced expression to the set
+            replaced_hashes.insert(common_expr.meta().hash);
+            // Create a new signal for the common subexpression
+            let (common_se, decomp) = create_common_ses_signal(&common_expr, &mut signal_factory);
+
+            // Add the new signal to the step type and a constraint for it
+            decomp.auto_signals.iter().for_each(|(q, expr)| {
+                if let Queriable::Internal(signal) = q {
+                    step_type_with_hash.add_internal(*signal);
+                }
+                step_type_with_hash.auto_signals.insert(*q, expr.clone());
+                step_type_with_hash.add_constr(format!("{:?}", q), expr.clone());
```

Comment:
Either expr is already q - common_ses and then autosignal should be q, common_ses or expr is common_ses and constraint should be 1-expr.

How is this not coming in the tests?

Reply:
Can you check now? I fixed one redundant constraint that was being added which was from this line: `step_type_with_hash.add_constr(format!("{:?}", q), expr.clone());`

And I added the explicit test to make sure we have the constraint that checks the equality of the auto_signal with the common subexpression, which in this case is common_ses - q, equivalent to what you said.

What is there right now is that when I create the new signal it is created on the `ConstrDecomp` because before this was being done for more than one common subexpr. I didn't remove the `ConstrDecomp` because I thought it would be needed later for the witness generation but I might be wrong. Anyways it seems a bit more modular this way and the auto_eq method is being reused for applying directly the constraint but if you think it causes confusion I can refactor in a way that the  `ConstrDecomp` is not used anymore. Lmk your thoughts pls

Reply:
It is not needed for witness generation, when you add it to auto_signals that makes the witness generation. `ConstrDecomp` is not stored anywhere so it is not needed for this. It could be use like before for more than one. Now it reduces the readability because it seems it is necessary for something when it is not. You are putting this info into the structure and then getting it out using map, which makes it less clear that there is only one, for example.

Reply:
Ok, so agreed and removed 👍🏼